server_addon/applications/server/__init__.py

```diff
@@ -129,6 +148,85 @@ async def create_applications_attribute(self) -> bool:
             bool: 'True' if an attribute was created or updated.
         """
 
+        need_restart = await self.create_required_attributes()
+        await self.update_enums()
+        return need_restart
+
+    async def create_required_attributes(self) -> bool:
+        """Make sure there are required 'applications' and 'tools' attributes.
+        This only checks for the existence of the attributes, it does not populate
+        them with any data. When an attribute is added, server needs to be restarted,
+        while adding enum data to the attribute does not require a restart.
+        Returns:
+            bool: 'True' if an attribute was created or updated.
+        """
+
+        # keep track of the last attribute position (for adding new attributes)
+        apps_present = False
+        tools_present = False
+        apps_attribute_data = self._get_applications_def()
+        apps_attrib_name = apps_attribute_data["name"]
+
+        tools_attribute_data = self._get_tools_def()
+        tools_attrib_name = tools_attribute_data["name"]
+
+        last_index = -1
+        async for row in Postgres.iterate(
+            "SELECT name, position FROM attributes ORDER BY position"
+        ):
+            # check if the required attributes are present
+            # (in that case, we don't need to add them)
+            # also keep track of the last attribute position
+            if row["name"] == apps_attrib_name:
+                apps_present = True
+            elif row["name"] == tools_attrib_name:
+                tools_present = True
+            last_index = row["position"]
```

Comment:
Wouldn't this query be much more optimal to include a filter where name is either of the two values instead of returning all attributes? For example adding `WHERE name IN ("applications", "tools")`?

We seem to be only doing this to get the highest available index? Shouldn't there be better ways to insert as highest index + 1 using the database calls themselves?

Reply:
I don't know about any other, @martastain ?

Reply:
whatever makes you happy:


```
SELECT MAX(position) AS max_position, BOOL_OR(name = 'applications') AS has_applications,  BOOL_OR(name = 'tools') AS has_tools FROM attributes;

 max_position | has_applications | has_tools 
--------------+------------------+-----------
           37 | t                | t
```